### ISSUE TITLE
[Bug] Effects Randomly get Destroyed

### ISSUE THREAD

**Issue by Gamingtoday093** [SmartlyDressedGames/Unturned-3.x-Community#3662]
So I've been developing a plugin that uses Effects as an indicator to players where to go to get to their marker but I've been encountering a weird inconsistent issue with effects.

***Sometimes*** all of the effects or a group of the effects refuse to spawn or get deleted/destroyed/despawn prematurely. This is not related to Lifetime since its grouped together and sometimes it works perfectly fine and all the effects last roughly the same amount of time. I don't believe its related to effect count since most of the time there won't be any issues sending around hundred effects.

![EffectsIssue](https://user-images.githubusercontent.com/70017362/214185674-d78855b2-8db3-49ff-b1fd-4d2d91a1a06e.gif)

![EffectsIssue2](https://user-images.githubusercontent.com/70017362/214185168-f7b570df-6095-4a73-bcb1-9cd32270ab01.gif)

I tried to look if this had been reported before with no avail so I hope this isn't a duplicate issue

**Comment by SDGNelson** [comment]
Are those two separate clips of the effect disappearing too soon?

What code is being used to spawn the effects?

Is it a vanilla effect? (looks like the airdrop one?)

Considering that non-UI effects don't have an option to destroy them from the server side what is the plugin using to control their lifetime?

**Comment by Gamingtoday093** [comment]
> Are those two separate clips of the effect disappearing too soon?

Yes, they are about a minute apart. Because it's really inconsistent.

> What code is being used to spawn the effects?

I call `EffectManager.ClearEffectByGuid(effectAsset.GUID, transportConnection);` to first clear all the effects.
Then in a loop foreach effect I want to spawn in I use this:
```cs
TriggerEffectParameters effectParameters = new TriggerEffectParameters(effectAsset);
effectParameters.reliable = true;
effectParameters.SetRelevantPlayer(transportConnection);
effectParameters.position = position;
effectParameters.direction = Vector3.forward;

EffectManager.triggerEffect(effectParameters);
```

> Is it a vanilla effect? (looks like the airdrop one?)

Yes, I am currently using the airdrop effect for testing. But that does not have an affect on anything. I have encountered this exact issue about a year ago using custom effects. (Custom effects with a really long Lifetime and no Lifetime  Spread)

> Considering that non-UI effects don't have an option to destroy them from the server side what is the plugin using to control their lifetime?

I clear them using `EffectManager.ClearEffectByGuid(effectAsset.GUID, transportConnection);` when I want to refresh the path indicators.

I am not too accustomed to Unturned's Networking solution but after thinking about this issue for some time yesterday I think I've come up with a somewhat good guess. Assuming that Unturned doesn't always recieve / execute messages from the server in order. I believe that `EffectManager.ClearEffectByGuid()` may be getting executed client side after already spawning in some of the effects. This would explain why it only seems to happen randomly once after spawning in effects. If this is the case, an easy solution would be to create a new `ClientStaticMethod` that specifically supports sending effect clear and multiple effects in one package. But ideally something should be done about this presumably unexpected behaviour.

**Comment by SDGNelson** [comment]
Oh right I forgot about that method! I suppose it works perfectly if the effect is only used for one thing.

Since they are both sent in reliable mode they will be executed in order. (`ClearEffectByGuid` is reliable too)

I'm not sure what else to suggest except the lifetime which you already mentioned it happens with that enabled.

Are you able to do a minimal reproducible example? For example a command that clears the effect and then spawns 100 of the airdrop effect? If it happens in that case I could run the same code and debug why it happens.

**Comment by Gamingtoday093** [comment]
> Are you able to do a minimal reproducible example? For example a command that clears the effect and then spawns 100 of the airdrop effect? If it happens in that case I could run the same code and debug why it happens.

Sure I'll try my best to get a reproducible example. But since its so inconsistent; it may have to do with something with *when* the effects are spawned in?, I'm not sure if I can get anything super reliable.

**Comment by Gamingtoday093** [comment]
Alright I've now spent a hell a lot of time trying to figure this out. And I believe I've actually found the problem.

Imagine if you will the lifetime of an Effect;
![lifetime](https://user-images.githubusercontent.com/70017362/215275060-18ef3f41-21df-4338-97e6-7334ce6db336.png)
The Blue represents the original effect's active duration. The Red represents the duration until the effect gets destroyed from lifetime. The Purple represents the same PoolReference Effect spawning in again after being cleared by GUID previously.

Now how is it that the original call to destroy Blue after a certain duration doesn't get cleared when it's cleared by GUID,
simply because of `TimeUtility.InvokeAfterDelay()` which triggers `InternalDestroyIntoPool()` which does not have any checks to make sure that the `InvokeAfterDelay` which triggered it was called when this effect was instantiated.

https://github.com/Unturned-Datamining/Unturned-Datamining/blob/599d03f85c1ad0537863ed7c7638e4532c3e1b8d/Assembly-CSharp/SDG.Unturned/PoolReference.cs#L34

`pool.Destroy()` only checks if the `PoolReference` isn't in the pool already which the Purple effect is not since its been spawned in. So it gets destroyed prematurely.
